src/cpp/session/modules/SessionRParser.cpp

```diff
@@ -0,0 +1,1287 @@
+          isWhitespace(cursor.nextToken()))
+      {
+         status.lint().unexpectedWhitespaceAroundOperator(cursor);
+      }
+      
+      // Extraction operators must be followed by a string or an
+      // identifier
+      if (isExtraction)
+      {
+         const RToken& next = cursor.nextSignificantToken();
+         if (!(next.isType(RToken::ID) || next.isType(RToken::STRING)))
+            status.lint().unexpectedToken(next);
+      }
+   }
+   
+   // There should be whitespace around other operators.
+   //
+   //    x+1
+   //
+   // is bad style.
+   else
+   {
+      if (!isWhitespace(cursor.previousToken()) ||
+          !isWhitespace(cursor.nextToken()))
+      {
+         status.lint().expectedWhitespaceAroundOperator(cursor);
+      }
+   }
+}
+
+void validateFunctionCall(const RTokenCursor& cursor,
+                          ParseStatus& status)
+{
+   if (!cursor.isType(RToken::LPAREN))
+      return;
+   
+   RTokenCursor startCursor = cursor.clone();
+   if (!startCursor.moveToPreviousSignificantToken())
+      return;
+   
+   if (!startCursor.moveToStartOfEvaluation())
+      return;
+   
+   RTokenCursor endCursor = cursor.clone();
+   if (!endCursor.fwdToMatchingToken())
+      return;
+
+   std::string fnNameString = string_utils::wideToUtf8(
+       std::wstring(startCursor.begin(), cursor.begin()));
+   
+   std::string fnCallString = string_utils::wideToUtf8(
+            std::wstring(startCursor.begin(), endCursor.end()));
+   
+   r::exec::RFunction validate(".rs.validateFunctionCall");
+   validate.addParam(fnNameString);
+   validate.addParam(fnCallString);
+   
+   std::string message;
+   Error error = validate.call(&message);
+   if (error)
+   {
+      LOG_ERROR(error);
+      return;
+   }
+   
+   if (!message.empty())
+   {
+      status.lint().add(
+               startCursor.row(),
+               startCursor.column(),
+               endCursor.row(),
+               endCursor.column() + endCursor.length(),
+               LintTypeError,
+               message);
+   }
+}
+
+bool skipFormulas(RTokenCursor& cursor,
+                  ParseStatus& status)
+{
+   if (cursor.nextSignificantToken().contentEquals(L"~"))
+      cursor.moveToNextSignificantToken();
+   
+   if (cursor.contentEquals(L"~"))
+   {
+      if (cursor.moveToEndOfStatement(status.isInParentheticalScope()))
+         return cursor.moveToNextSignificantToken();
+      
+   }
+   
+   return false;
+}
+
+} // anonymous namespace
+
+#define GOTO_INVALID_TOKEN(__CURSOR__)                                         \
+   do                                                                          \
+   {                                                                           \
+      DEBUG("Invalid token: " << __CURSOR__);                                  \
+      goto INVALID_TOKEN;                                                      \
+   } while (0)
+
+void doParse(RTokenCursor& cursor,
+             ParseStatus& status)
```

Comment:
By (file-local) convention I decided to go ahead and just pass these parameters by reference, rather than pointer -- since all methods within that take the cursor and status will potentially mutate them.

I think in this small case it's well-understood that each method will (potentially) modify each reference passed in, but I can certainly switch to using pointers if preferred.